Introduction of SingleToMultiLinePhpdocFixer

BackEndTea opened the pull request:
Fixes: #2646 

If preferred, i can target the changes to the Doc block  class to the current LTS (2.2?), as its features are useful for a number of other fixers. 

```bash
$ ./php-cs-fixer describe single_to_multi_line_phpdoc
Description of single_to_multi_line_phpdoc rule.
Changes doc blocks from single to multi line, or reversed.

Fixer is configurable using following options:
* const ('multi', 'single'): whether const blocks should be single or multi line; defaults to 'multi'
* property ('multi', 'single'): whether property doc blocks should be single or multi line; defaults to 'multi'
* method ('multi', 'single'): whether method doc blocks should be single or multi line; defaults to 'multi'
```
Fixing examples:
 * Example #1. Fixing with the default configuration.
   ---------- begin diff ----------
```diff
<?php
class Foo{
-    /** @var bool */
+    /**
+     * @var bool
+     */
    public $var;
}
```
   ----------- end diff -----------

 * Example #2. Fixing with configuration: ['property' => 'single'].
   ---------- begin diff ----------
```diff
<?php
class Foo{
-    /**
-    * @var bool
-    */
+    /** @var bool */
    public $var;
 }
```
   ----------- end diff -----------

SpacePossum commented:
Nice work and :+1:  for concept.
About priorities there is some stuff to check out I think, like when rules remove lines (`phpdoc_no_useless_inheritdoc` and others) or add lines (`phpdoc_add_missing_param_annotation` and possibly others), together with locking down the priority with `phpdoc_to_comment`

ntzm commented on tests/Fixtures/Integration/priority/general_phpdoc_annotation_remove,single_to_multi_line_phpdoc.test:
```diff
@@ -0,0 +1,28 @@
+--TEST--
+Integration of fixers: general_phpdoc_annotation_remove,single_to_multi_line_phpdoc.
+--RULESET--
+{"general_phpdoc_annotation_remove": {"annotations": ["test"] }, "single_to_multi_line_phpdoc": {"method": "single"}}
```
extra space after `["test"]`

keradus replied:
done

ntzm commented on src/Fixer/Phpdoc/SingleToMultiLinePhpdocFixer.php:
```diff
@@ -0,0 +1,144 @@
+    {
+        do {
+            $index = $tokens->getPrevNonWhitespace($index);
+        } while ($tokens[$index]->isGivenKind([T_PUBLIC, T_PROTECTED, T_PRIVATE, T_FINAL, T_ABSTRACT, T_COMMENT]));
```
what about `T_VAR`?

dmvdbrugge replied:
And what about `T_STATIC`?

hkdobrev commented:
@BackEndTea Do you think that could be finalised for 2.14?

BackEndTea commented:
I won't have time until somewhere next week to work on this.

TomasVotruba commented:
@BackEndTea Thanks for your work, this is awesome feature :+1:

keradus commented:
> I won't have time until somewhere next week to work on this.

@BackEndTea , feel free to ping us when you consider this ready.
in the meantime, i'm changing the milestone for this PR

BackEndTea commented:
@keradus Sorry for the long delay on this, I kinda forgot about it.

Currently this only works for class methods, properties and constants, and i'm not 100% sure how to expand it to also work with functions, classes and other doc blocks.

jaikdean commented:
I agree with SpacePossum, it would be great to see this released as it stands.

BackEndTea commented:
@SpacePossum & @keradus  Sorry for the long wait on this, but I have updated it, and it is ready for review.

I have rebased this and fixed a few remarks people had.
Current concerns are:
* The fact that its 'incomplete' e.g. only class properties, constants and methods. But this should work as is.
* Priorities with other fixes
* The fact that this includes updates to the `DocBlock` class. This isn't a final class and could be extended by others.

keradus commented:
@BackEndTea ,
the CI is failing, this is a blocker to proceed with the PR :(
can you fix it, please ?

staabm commented:
@BackEndTea do you have time to work on this?

gharlan commented:
I think we should find a better name for this one, because it can be used for both directions.

dmvdbrugge commented:
> I think we should find a better name for this one, because it can be used for both directions.

I think replacing `to` with `or` makes clear that it's purpose is to make sure it's always either single or multi, so `SingleOrMultiLinePhpdocFixer`. In fact, I see now that the branch is already called that 😆

localheinz commented:
Perhaps `PhpdocLineSpanFixer`?

SpacePossum commented:
rebased, conflicts resolved, fixed unit tests, fixed code style to new standard, rename the fixer, forced pushed, in preparation of merging

SpacePossum commented:
Thank you @BackEndTea.

TomasVotruba commented:
Thank you @BackEndTea, gonna deprecate my own rule now :)
